### PR TITLE
IRGen: Lookup the conformance of an archetype in the right substitution map when computing necessary bindings of abstract cnditional requirements

### PR DESCRIPTION
**Pull request by aschwaighofer**
SR-12853

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -2749,9 +2749,12 @@ static void addAbstractConditionalRequirements(
     auto *proto =
         req.getSecondType()->castTo<ProtocolType>()->getDecl();
     auto ty = req.getFirstType()->getCanonicalType();
-    if (!isa<ArchetypeType>(ty))
+    auto archetype =  dyn_cast<ArchetypeType>(ty);
```
Comment (slavapestov) [inline lib/IRGen/GenProto.cpp:2752]: Nitpick: extra space

File: lib/IRGen/GenProto.cpp [hunk 2]
```diff
@@ -2749,9 +2749,12 @@ static void addAbstractConditionalRequirements(
...
+    if (!archetype)
       continue;
-    auto conformance = subMap.lookupConformance(ty, proto);
+    auto *genericEnv = archetype->getGenericEnvironment();
+    auto conformance =
+        genericEnv->getForwardingSubstitutionMap().lookupConformance(ty, proto);
```
Comment (slavapestov) [inline lib/IRGen/GenProto.cpp:2757]: Is this ever not just an abstract conformance? Look at how the forwarding substitution map is constructed...

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Pull request merged**